pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -375,8 +375,17 @@ func (ssc *defaultStatefulSetControl) processReplica(
 	replicas []*v1.Pod,
 	i int) (bool, error) {
 	logger := klog.FromContext(ctx)
-	// delete and recreate failed pods
-	if isFailed(replicas[i]) {
+	// Delete and recreate pods which finished running.
+	//
+	// Note that pods with phase Succeeded will also trigger this event. This is
+	// because final pod phase of evicted or otherwise forcibly stopped pods
+	// (e.g. terminated on node reboot) is determined by the exit code of the
+	// container, not by the reason for pod termination. We should restart the pod
+	// regardless of the exit code.
+	//
+	// Also note that it is impossible for StatefulSet pods to ever truly complete,
+	// as validation allows only restartPolicy=Always.
```

Comment:
I would remove this last paragraph. It is confusing, as we can see that it will actually complete in case of VM preemption or node pressure.